### ISSUE TITLE
Inductor constant-folds `sum(softmax)` to exactly 1.0, causing `floor()` to produce wrong integer results

### ISSUE THREAD

**Issue by XYZboom** [pytorch/pytorch#189787]
### 🐛 Describe the bug

`torch.compile` with the **Inductor** backend constant-folds `sum(softmax(x, dim))` to exactly `1.0` (exploiting the mathematical identity). However, in finite-precision float32 arithmetic, softmax sums can be slightly below or above 1.0 due to rounding. When `floor()` is applied downstream, this creates a **0 vs 1** discrepancy between eager and compiled modes.

Using `torch._dynamo.utils.same` with an fp64 reference, the raw `sum(softmax)` outputs from both eager and compile **pass** the comparison — they are both within expected float32 tolerance of fp64. However, `floor()` amplifies this ULP-level difference into an integer-level mismatch, and **Inductor's constant-folded `1.0` is actually further from the fp64 ground truth than the eager result** in some cases.

## Minimal reproducer

```python
import torch
import torch.nn.functional as F
import torch._dynamo
import torch._inductor.config

# Fuzzer requirement: fallback_random + seed reset
torch._inductor.config.fallback_random = True

torch.manual_seed(42)
x = torch.randn(43, 59, dtype=torch.float32, device="cpu")
x_fp64 = x.to(torch.float64)

class SubMod(torch.nn.Module):
    def forward(self, x):
        s = F.softmax(x, dim=-1)
        total = torch.sum(s, dim=-1, keepdim=False)
        return torch.floor(total)

model = SubMod()

# Eager (with seed reset)
torch.manual_seed(42)
with torch.no_grad():
    eager = model(x)

# Compile (with seed reset)
torch._dynamo.reset()
torch.manual_seed(42)
compiled = torch.compile(model, mode="default")
with torch.no_grad():
    comp = compiled(x)

# Verify raw sums pass torch._dynamo.utils.same with fp64 reference
class SubModRaw(torch.nn.Module):
    def forward(self, x):
        s = F.softmax(x, dim=-1)
        total = torch.sum(s, dim=-1, keepdim=False)
        return total

model_raw = SubModRaw()
with torch.no_grad():
    fp64_ref = model_raw(x_fp64)
    fp32_eager_raw = model_raw(x)

torch._dynamo.reset()
compiled_raw = torch.compile(model_raw, mode="default")
with torch.no_grad():
    fp32_comp_raw = compiled_raw(x)

same_result = torch._dynamo.utils.same(fp32_eager_raw, fp32_comp_raw, fp64_ref=fp64_ref)
print(f"torch._dynamo.utils.same (raw sums, fp64 ref): {same_result}")

# Show floor mismatch
diff_idx = (eager != comp)
print(f"eager:   {eager[:10].tolist()}")
print(f"compile: {comp[:10].tolist()}")
print(f"floor() mismatch: {diff_idx.sum().item()} / {eager.numel()}")

# fp64 ground truth
with torch.no_grad():
    fp64_floor = torch.floor(model_raw(x_fp64))
comp_missed = ((fp64_floor == 0) & (comp == 1)).sum().item()
print(f"compile wrong per fp64 (returns 1, should be 0): {comp_missed}")
```

### Output

```
torch._dynamo.utils.same (raw sums, fp64 ref): True
eager:   [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0]
compile: [1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 1.0]
floor() mismatch: 13 / 43
compile wrong per fp64 (returns 1, should be 0): 11
```

### Error logs

No error is raised — this is a **silent correctness** issue. The compiled model produces different integer results from eager mode without any warning or exception.

### Versions

Collecting environment information...
PyTorch version: 2.13.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 15.0.7
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.15 (main, Jun 11 2026, 15:20:16) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-35-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               AuthenticAMD
Model name:                              AMD Ryzen 9 9950X 16-Core Processor
CPU family:                              26
Model:                                   68
Thread(s) per core:                      2
Core(s) per socket:                      16
Socket(s):                               1
Stepping:                                0
Frequency boost:                         enabled
CPU(s) scaling MHz:                      69%
CPU max MHz:                             5756.4521
CPU min MHz:                             624.1940
BogoMIPS:                                8599.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpuid_fault cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx_vnni avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid bus_lock_detect movdiri movdir64b overflow_recov succor smca fsrm avx512_vp2intersect flush_l1d amd_lbr_pmc_freeze
Virtualization:                          AMD-V
L1d cache:                               768 KiB (16 instances)
L1i cache:                               512 KiB (16 instances)
L2 cache:                                16 MiB (16 instances)
L3 cache:                                64 MiB (2 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; IBPB on VMEXIT only
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB on VMEXIT

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvtx==13.0.85
[pip3] torch==2.13.0
[pip3] triton==3.7.1
[conda] numpy                     2.4.6                    pypi_0    pypi
[conda] nvidia-cublas             13.1.1.3                 pypi_0    pypi
[conda] nvidia-cuda-cupti         13.0.85                  pypi_0    pypi
[conda] nvidia-cuda-nvrtc         13.0.88                  pypi_0    pypi
[conda] nvidia-cuda-runtime       13.0.96                  pypi_0    pypi
[conda] nvidia-cudnn-cu13         9.20.0.48                pypi_0    pypi
[conda] nvidia-cufft              12.0.0.61                pypi_0    pypi
[conda] nvidia-curand             10.4.0.35                pypi_0    pypi
[conda] nvidia-cusolver           12.0.4.66                pypi_0    pypi
[conda] nvidia-cusparse           12.6.3.3                 pypi_0    pypi
[conda] nvidia-cusparselt-cu13    0.8.1                    pypi_0    pypi
[conda] nvidia-nccl-cu13          2.29.7                   pypi_0    pypi
[conda] nvidia-nvjitlink          13.3.33                  pypi_0    pypi
[conda] nvidia-nvtx               13.0.85                  pypi_0    pypi
[conda] torch                     2.13.0                   pypi_0    pypi
[conda] triton                    3.7.1                    pypi_0    pypi

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by XYZboom** [comment]
Sorry, I forget to add the label `fuzzer`.

**Comment by aorenste** [comment]
Slightly simpler repro:
```
import torch
import torch.nn.functional as F

x = torch.tensor(
    [[-0.2763414680957794, -0.19321373105049133, 0.7767809629440308, 0.6838752627372742, -1.3245893716812134,
      -0.5160817503929138, 0.6001842617988586, -0.4702208340167999, -0.6086435317993164, -0.046192023903131485,
      -1.6457397937774658, -0.4833274185657501, -0.740294337272644, 0.31428107619285583, 0.1415553092956543,
      1.0348176956176758, -0.626437783241272, -0.5150922536849976, 0.6902899742126465, -0.4939993619918823,
      1.1366126537322998, -0.46184006333351135, 1.419979453086853, 0.848518967628479, -0.047891248017549515,
      0.6685602068901062, 1.0429801940917969, 0.6899017095565796, -1.3129348754882812, 0.03780364990234375,
      -1.1702114343643188, -0.10318559408187866, 1.1894739866256714, 0.7606944441795349, -0.7463049292564392,
      -1.3838845491409302, 0.4868715703487396, -1.0020296573638916, 0.0329488180577755, -0.42919591069221497,
      -0.9817978739738464, -0.6420586109161377, 0.8265887498855591, 1.591395616531372, -0.1208132952451706,
      -0.48302069306373596, 0.11329790204763412, 0.0771508663892746, -0.9228128790855408, -1.2619991302490234,
      1.0860532522201538, 1.096641182899475, -0.6836934685707092, 0.06604336202144623, -0.0007737990817986429,
      0.1620604395866394, 1.195958137512207, -1.3061535358428955, -1.4039719104766846]],
    dtype=torch.float32,
    device="cpu")

class SubMod(torch.nn.Module):
    def forward(self, x):
        s = F.softmax(x, dim=-1)
        total = torch.sum(s, dim=-1, keepdim=False)
        return total
        return torch.floor(total)

model = SubMod()

with torch.no_grad():
    eager = model(x)

compiled = torch.compile(model, backend="inductor")
with torch.no_grad():
    comp = compiled(x)

import struct
print(f"eager:   {eager.tolist()[0]}")
print(f"compile: {comp.tolist()[0]}")
```

**Comment by aorenste** [comment]
This is a numerics accuracy issue amplified by the floor. Leaving to the inductor team to prioritize.
